core/src/main/java/com/radixdlt/genesis/olympia/converter/resource/OlympiaNonXrdConverter.java

```diff
@@ -0,0 +1,309 @@
+    for (var balance : olympiaNonXrdBalances) {
+      final var resIdx = balance.resourceIndex();
+      if (resourceTotalSuppliesOnOlympia[resIdx] == null) {
+        resourceTotalSuppliesOnOlympia[resIdx] = balance.amount();
+      } else {
+        resourceTotalSuppliesOnOlympia[resIdx] =
+            resourceTotalSuppliesOnOlympia[resIdx].add(balance.amount());
+      }
+    }
+    return resourceTotalSuppliesOnOlympia;
+  }
+
+  private static Decimal scaleResourceAmount(
+      BigInteger originalAmount,
+      BigInteger resourceTotalSupplyOnOlympia,
+      BigInteger resourceMaxSupplyOnBabylon) {
+    if (resourceTotalSupplyOnOlympia.compareTo(resourceMaxSupplyOnBabylon) <= 0) {
+      // No need to scale, guaranteed to fit
+      return Decimal.unsafeFromRawBigIntRepr(originalAmount);
+    } else {
+      // Scale it down, using integer div rounding
+      final var scaledBigInt =
+          resourceMaxSupplyOnBabylon.multiply(originalAmount).divide(resourceTotalSupplyOnOlympia);
+      return Decimal.unsafeFromRawBigIntRepr(scaledBigInt);
+    }
+  }
+
+  private static ImmutableList<GenesisDataChunk.Resources> prepareResourcesChunks(
+      OlympiaToBabylonConverterConfig config,
+      List<OlympiaStateIR.Account> accounts,
+      List<OlympiaStateIR.Resource> resources,
+      int olympiaXrdResourceIndex,
+      Decimal[] resourceTotalSuppliesOnBabylon) {
+    return createChunks(
+        resources,
+        config.maxResourcesPerChunk(),
+        (idx, olympiaResource) -> {
+          if (idx == olympiaXrdResourceIndex) {
+            // skip XRD
+            return Optional.empty();
+          }
+          final var initialSupply =
+              resourceTotalSuppliesOnBabylon[idx] == null
+                  ? Decimal.ZERO
+                  : resourceTotalSuppliesOnBabylon[idx];
+          return Optional.of(convertResource(accounts, initialSupply, olympiaResource));
+        },
+        GenesisDataChunk.Resources::new);
+  }
+
+  private static GenesisResource convertResource(
+      List<OlympiaStateIR.Account> accounts,
+      Decimal initialSupply,
+      OlympiaStateIR.Resource resource) {
+    final var metadata =
+        ImmutableList.of(
+            tuple("symbol", resource.symbol()),
+            tuple("name", resource.name()),
+            tuple("description", resource.description()),
+            tuple("url", resource.url()),
+            tuple("icon_url", resource.iconUrl()));
+    final var owner =
+        resource
+            .ownerAccountIndex()
+            .map(
+                idx -> Address.virtualAccountAddress(accounts.get(idx).publicKeyBytes().asBytes()));
+
+    final var addrBytes = padToResourceAddress(resource.addr().getBytes());
+    return new GenesisResource(addrBytes, initialSupply, metadata, Option.from(owner));
+  }
+
+  private static byte[] padToResourceAddress(byte[] input) {
+    return Bytes.leftPadWithZeros(input, ResourceAddress.BYTE_LENGTH - 1);
```

Comment:
Hmmm I don't like that these addresses are going to all look special (kinda like vanity addresses) - which is detrimental to the system addresses looking special. And also the `q`s are ugly.

Could we instead just hash the bytes? It's still deterministic, just not invertible. (Just like the account mapping).

Reply:
:+1: